src/psyclone/psyir/transformations/psy_data_trans.py

```diff
@@ -0,0 +1,192 @@
+
+        # The checks below are only for the NEMO API and can be removed
+        # once #435 is done.
+        invoke = node_list[0].root.invoke
+        if not isinstance(invoke, NemoInvoke):
+            return
+
+        # Get the parse tree of the routine containing this region
+        # pylint: disable=protected-access
+        ptree = invoke._ast
+        # pylint: enable=protected-access
+        # Search for the Specification_Part
+        if not walk([ptree], Fortran2003.Specification_Part):
+            raise TransformationError(
+                "For the NEMO API, PSyData can only be added to routines "
+                "which contain existing variable declarations (i.e. a "
+                "Specification Part) but '{0}' does not have any.".format(
+                    invoke.name))
+
+    def apply(self, nodes, options=None):
+        # pylint: disable=arguments-differ
+        '''Apply this transformation to a subset of the nodes within a
+        schedule - i.e. enclose the specified Nodes in the
+        schedule within a single PSyData region.
+
+        :param nodes: can be a single node or a list of nodes.
+        :type nodes: :py:obj:`psyclone.psygen.Node` or list of \
+                     :py:obj:`psyclone.psygen.Node`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :returns: Tuple of the modified schedule and a record of the \
+                  transformation.
+        :rtype: (:py:class:`psyclone.psyGen.Schedule`, \
+                :py:class:`psyclone.undoredo.Memento`)
+
+        '''
+        # Check whether we've been passed a list of nodes or just a
+        # single node.
+        if isinstance(nodes, list) and isinstance(nodes[0], Node):
+            node_list = nodes
+        elif isinstance(nodes, Schedule):
+            # We've been passed a Schedule so default to enclosing its
+            # children.
+            node_list = nodes.children
+        elif isinstance(nodes, Node):
+            # Single node that's not a Schedule
+            node_list = [nodes]
+        else:
+            arg_type = str(type(nodes))
+            raise TransformationError("Error in {1}. "
+                                      "Argument must be a single Node in a "
+                                      "schedule or a list of Nodes in a "
+                                      "schedule but have been passed an "
+                                      "object of type: {0}".
+                                      format(arg_type, str(self)))
+
+        # Perform validation checks
+        self.validate(node_list, options)
+
+        # create a memento of the schedule and the proposed
+        # transformation
+        schedule = node_list[0].root
+        keep = Memento(schedule, self)
+
+        # Keep a reference to the parent of the nodes that are to be
+        # enclosed within a profile region. Also keep the index of
```

Comment:
s/profile/PSyData/ and s/Profile/PSyData/ below.

Reply:
Done.